module/zfs/zfs_ioctl.c

```diff
@@ -2819,8 +2777,49 @@ zfs_ioc_inherit_prop(zfs_cmd_t *zc)
 			return (SET_ERROR(EINVAL));
 	}
 
-	/* property name has been validated by zfs_secpolicy_inherit_prop() */
-	return (dsl_prop_inherit(zc->zc_name, zc->zc_value, source));
+	if (prop == ZPROP_INVAL) {
+		if (!zfs_prop_user(propname))
+			return (SET_ERROR(EINVAL));
+
+		type = PROP_TYPE_STRING;
+	} else if (prop == ZFS_PROP_VOLSIZE || prop == ZFS_PROP_VERSION) {
+		return (SET_ERROR(EINVAL));
+	} else {
+		type = zfs_prop_get_type(prop);
+	}
+
+	/*
+	 * zfs_prop_set_special() expects properties in the form of an
+	 * nvpair with type info.
+	 */
+	dummy = fnvlist_alloc();
+
+	switch (type) {
+	case PROP_TYPE_STRING:
+		VERIFY(0 == nvlist_add_string(dummy, propname, ""));
+		break;
+	case PROP_TYPE_NUMBER:
+	case PROP_TYPE_INDEX:
+		VERIFY(0 == nvlist_add_uint64(dummy, propname, 0));
+		break;
+	default:
+		err = SET_ERROR(EINVAL);
+		goto errout;
+	}
+
+	pair = nvlist_next_nvpair(dummy, NULL);
+	if (pair == NULL) {
+		err = SET_ERROR(EINVAL);
+	} else {
+		err = zfs_prop_set_special(zc->zc_name, source, pair);
+		if (err == -1) /* property is not "special", needs handling */
+			err = dsl_prop_inherit(zc->zc_name, zc->zc_value,
```

Comment:
> In case of snapdev however, setting the top-level filesystem property results in the values propagated but the source is 'local', not 'inherited from ...'. Also, regardless of explicit setting of snapdev in the middle of the hierarchy, the changes to top-level fs are always propagated down the hierarchy, which is not quite how this works for dedup.

> If it helps, I can attach a terminal session log that illustrates the points made above.

Please do, my testing is limited but my changes seem to be working as intended:

```
root@zfstest:~# # ensure we're running the correct zfs version
root@zfstest:~# cat /sys/module/zfs/version
0.7.0-rc4_39_g5fc10596e
root@zfstest:~# #
root@zfstest:~# # setup
root@zfstest:~# POOLNAME='testpool'
root@zfstest:~# TMPDIR='/var/tmp'
root@zfstest:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@zfstest:~# zpool destroy $POOLNAME
root@zfstest:~# rm -f $TMPDIR/zpool.dat
root@zfstest:~# fallocate -l 128m $TMPDIR/zpool.dat
root@zfstest:~# zpool create -O mountpoint=none $POOLNAME $TMPDIR/zpool.dat
root@zfstest:~# zfs create $POOLNAME/fs1
root@zfstest:~# zfs create $POOLNAME/fs2
root@zfstest:~# zfs create -V 1M -s $POOLNAME/fs1/zvol11
root@zfstest:~# zfs create -V 1M -s $POOLNAME/fs1/zvol12
root@zfstest:~# zfs create -V 1M -s $POOLNAME/fs2/zvol21
root@zfstest:~# #
root@zfstest:~# # everything is default
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   hidden   -         default
testpool/fs1         snapdev   hidden   -         default
testpool/fs1/zvol11  snapdev   hidden   -         default
testpool/fs1/zvol12  snapdev   hidden   -         default
testpool/fs2         snapdev   hidden   -         default
testpool/fs2/zvol21  snapdev   hidden   -         default
root@zfstest:~# #
root@zfstest:~# # every dataset inherit from top-level
root@zfstest:~# zfs set snapdev=visible $POOLNAME
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   visible  -         local
testpool/fs1         snapdev   visible  -         inherited from testpool
testpool/fs1/zvol11  snapdev   visible  -         inherited from testpool
testpool/fs1/zvol12  snapdev   visible  -         inherited from testpool
testpool/fs2         snapdev   visible  -         inherited from testpool
testpool/fs2/zvol21  snapdev   visible  -         inherited from testpool
root@zfstest:~# #
root@zfstest:~# # fs1 is 'local', its children inherit. everything else is same as before
root@zfstest:~# zfs set snapdev=hidden $POOLNAME/fs1
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   visible  -         local
testpool/fs1         snapdev   hidden   -         local
testpool/fs1/zvol11  snapdev   hidden   -         inherited from testpool/fs1
testpool/fs1/zvol12  snapdev   hidden   -         inherited from testpool/fs1
testpool/fs2         snapdev   visible  -         inherited from testpool
testpool/fs2/zvol21  snapdev   visible  -         inherited from testpool
root@zfstest:~# #
root@zfstest:~# # fs1 will now inherit from the top-level, again
root@zfstest:~# zfs inherit snapdev $POOLNAME/fs1
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   visible  -         local
testpool/fs1         snapdev   visible  -         inherited from testpool
testpool/fs1/zvol11  snapdev   visible  -         inherited from testpool
testpool/fs1/zvol12  snapdev   visible  -         inherited from testpool
testpool/fs2         snapdev   visible  -         inherited from testpool
testpool/fs2/zvol21  snapdev   visible  -         inherited from testpool
```